python/pyspark/sql/session.py

```diff
@@ -231,6 +231,7 @@ def __init__(self, sparkContext, jsparkSession=None):
                 or SparkSession._instantiatedSession._sc._jsc is None:
             SparkSession._instantiatedSession = self
             self._jvm.SparkSession.setDefaultSession(self._jsparkSession)
+            self._jvm.SparkSession.setActiveSession(self._jsparkSession)
```

Comment:
@huaxingao, can you check if the active session is set? for instance when we `createDataFrame`? From a cursory look, we are not setting it.

Reply:
@HyukjinKwon Seems to me that active session is set OK in the ```__init__```. When createDataFrame, we already have a session, and the active session is already set in the ```__init__```.

Reply:
> When createDataFrame, we already have a session

but wouldn't we not set the active session properly if session A sets an active session in `__init__`, and then session B sets an active session in `__init__`, and then session A calls `createDataFrame` ?

Reply:
Simialr. I was expecting something like:

```python
session1 = SparkSession.builder.config("key1", "value1").getOrCreate()
session2 = SparkSession.builder.config("key2", "value2").getOrCreate()

assert(session2 == SparkSession.getActiveSession())

session1.createDataFrame([(1, 'Alice')], ['age', 'name'])

assert(session1 == SparkSession.getActiveSession())
```

does this work?

Reply:
So @HyukjinKwon in this code session1 and session2 are already equal:

> Welcome to
>       ____              __
>      / __/__  ___ _____/ /__
>     _\ \/ _ \/ _ `/ __/  '_/
>    /__ / .__/\_,_/_/ /_/\_\   version 2.3.1
>       /_/
> 
> Using Python version 3.6.5 (default, Apr 29 2018 16:14:56)
> SparkSession available as 'spark'.
> >>> session1 = SparkSession.builder.config("key1", "value1").getOrCreate()
> >>> session2 = SparkSession.builder.config("key2", "value2").getOrCreate()
> >>> session1
> <pyspark.sql.session.SparkSession object at 0x7ff6d4843b00>
> >>> session2
> <pyspark.sql.session.SparkSession object at 0x7ff6d4843b00>
> >>> session1 == session2
> True
> >>> 
> 
> 
> 
> 
> 

That being said the possibility of having multiple Spark session in Python is doable you manually have to call the init e.g.:

> >>> session3 = SparkSession(sc)
> >>> session3
> <pyspark.sql.session.SparkSession object at 0x7ff6d3dbd160>
> >>> 
> 

And supporting that is reasonable.

Reply:
If we're going to support this we should have test for it, or if we aren't going to support this right now we should document the behaviour.

Reply:
Oh, okay. I had to be explicit. I meant:

```scala
scala> import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.SparkSession

scala> SparkSession.getActiveSession
res0: Option[org.apache.spark.sql.SparkSession] = Some(org.apache.spark.sql.SparkSession@3ef4a8fb)

scala> val session1 = spark
session1: org.apache.spark.sql.SparkSession = org.apache.spark.sql.SparkSession@3ef4a8fb

scala> val session2 = spark.newSession()
session2: org.apache.spark.sql.SparkSession = org.apache.spark.sql.SparkSession@4b74a4d

scala> SparkSession.getActiveSession
res1: Option[org.apache.spark.sql.SparkSession] = Some(org.apache.spark.sql.SparkSession@3ef4a8fb)

scala> session2.createDataFrame(Seq(Tuple1(1)))
res2: org.apache.spark.sql.DataFrame = [_1: int]

scala> SparkSession.getActiveSession
res3: Option[org.apache.spark.sql.SparkSession] = Some(org.apache.spark.sql.SparkSession@4b74a4d)
```

Reply:
@holdenk @HyukjinKwon 
Thanks for the comments. I looked the scala code, it ```setActiveSession``` in ```createDataFrame```. 
```
  def createDataFrame[A <: Product : TypeTag](rdd: RDD[A]): DataFrame = {
    SparkSession.setActiveSession(this)
    ...
  }
```
I will do the same for python.
```
def createDataFrame(self, data, schema=None, samplingRatio=None, verifySchema=True):
        SparkSession._activeSession = self
        self._jvm.SparkSession.setActiveSession(self._jsparkSession)
```
Will also add a test